src/backend/distributed/planner/shard_pruning.c

```diff
@@ -578,25 +568,104 @@ PrunableExpressionsWalker(Node *node, ClauseWalkerContext *context)
 
 		return false;
 	}
-	else
+
+	return expression_tree_walker(node, PrunableExpressionsWalker, context);
+}
+
+
+/*
+ * AddSAOPartitionKeyRestrictionToInstance adds partcol = arrayelem operator
+ * restriction to the current pruning instance.
+ */
+static void
+AddSAOPartitionKeyRestrictionToInstance(ClauseWalkerContext *context,
+										ScalarArrayOpExpr *arrayOperatorExpression)
+{
+	PruningInstance *prune = context->currentPruningInstance;
+	Node *leftOpExpression = linitial(arrayOperatorExpression->args);
+	Node *strippedLeftOpExpression = strip_implicit_coercions(leftOpExpression);
+	bool usingEqualityOperator = OperatorImplementsEquality(
+		arrayOperatorExpression->opno);
+	Expr *arrayArgument = (Expr *) lsecond(arrayOperatorExpression->args);
+
+	/* found partcol = ANY(const, value, s); or partcol IN (const,b,c); */
+	if (usingEqualityOperator && strippedLeftOpExpression != NULL &&
+		equal(strippedLeftOpExpression, context->partitionColumn) &&
+		IsA(arrayArgument, Const))
 	{
-		PruningInstance *prune = context->currentPruningInstance;
+		ArrayType *array = NULL;
+		int16 typlen = 0;
+		bool typbyval = false;
+		char typalign = '\0';
+		Oid element_type = 0;
+		ArrayIterator arrayIterator = NULL;
+		Datum arg = 0;
+		Datum arrayDatum = ((Const *) arrayArgument)->constvalue;
+		bool isNull = false;
+
+		/* check for the NULL right-hand expression*/
+		if (arrayDatum == 0)
+		{
+			return;
+		}
+
+		array = DatumGetArrayTypeP(((Const *) arrayArgument)->constvalue);
 
+		/* get the necessary information from array type to iterate over it */
+		element_type = ARR_ELEMTYPE(array);
+		get_typlenbyvalalign(element_type,
+							 &typlen,
+							 &typbyval,
+							 &typalign);
+
+		/* Iterate over the righthand array of expression */
+		arrayIterator = array_create_iterator(array, 0, NULL);
+		while (array_iterate(arrayIterator, &arg, &isNull))
+		{
+			OpExpr *op = NULL;
+			Const *c = makeConst(element_type, -1,
+								 DEFAULT_COLLATION_OID,
+								 typlen,
+								 arg,
+								 isNull,
+								 typbyval);
+			PendingPruningInstance *instance = palloc0(sizeof(PendingPruningInstance));
+
+			/* build partcol = arrayelem operator */
+			op = makeNode(OpExpr);
+			op->opno = arrayOperatorExpression->opno;
+			op->opfuncid = arrayOperatorExpression->opfuncid;
+			op->inputcollid = arrayOperatorExpression->inputcollid;
+			op->opresulttype = get_func_rettype(arrayOperatorExpression->opfuncid);
+			op->opcollid = DEFAULT_COLLATION_OID;
+			op->location = -1;
+			op->args = list_make2(strippedLeftOpExpression, c);
```

Comment:
I need to pass all the info to create const element within this function. I may also implement array iterator part as a function, but I suspect that it will be used by another part of the code except here.